src/filters/oauth_filter.rs

```diff
@@ -0,0 +1,89 @@
+use crate::services::external_identity_validator::{
+    ExternalIdentityValidator, ExternalIdentityValidatorImpl,
+};
+use actix_web::dev::{forward_ready, Service, ServiceRequest, ServiceResponse, Transform};
+use actix_web::error::ErrorUnauthorized;
+use actix_web::Error;
+use futures_util::future::LocalBoxFuture;
+use std::future::{ready, Ready};
+use std::rc::Rc;
+
+// Middleware for external token validation factory
+pub struct ExternalTokenMiddlewareFactory {}
+
+// The ExternalTokenMiddlewareFactory's own methods implementation
+impl ExternalTokenMiddlewareFactory {
+    pub(crate) fn new() -> Self {
+        ExternalTokenMiddlewareFactory {}
+    }
+
+    pub(crate) fn create(&self) -> Rc<dyn ExternalIdentityValidator> {
+        Rc::new(ExternalIdentityValidatorImpl {})
+    }
+}
+
+// Transform trait implementation
+// `NextServiceType` - type of the next service
+// `BodyType` - type of response's body
+impl<NextService, BodyType> Transform<NextService, ServiceRequest>
+    for ExternalTokenMiddlewareFactory
+where
+    NextService:
+        Service<ServiceRequest, Response = ServiceResponse<BodyType>, Error = Error> + 'static,
+    NextService::Future: 'static,
+    BodyType: 'static,
+{
+    type Response = ServiceResponse<BodyType>;
+    type Error = Error;
+    type Transform = OauthMiddleware<NextService>;
+    type InitError = ();
+    type Future = Ready<Result<Self::Transform, Self::InitError>>;
+
+    fn new_transform(&self, service: NextService) -> Self::Future {
+        let validator = self.create();
+        let mw = OauthMiddleware {
+            service: Rc::new(service),
+            external_identity_validator: validator,
+        };
+        ready(Ok(mw))
+    }
+}
+
+// The middleware object
+pub struct OauthMiddleware<NextService> {
+    service: Rc<NextService>,
+    external_identity_validator: Rc<dyn ExternalIdentityValidator>,
+}
+
+// The middleware implementation
+impl<NextService, BodyType> Service<ServiceRequest> for OauthMiddleware<NextService>
+where
+    NextService:
+        Service<ServiceRequest, Response = ServiceResponse<BodyType>, Error = Error> + 'static,
+    NextService::Future: 'static,
+    BodyType: 'static,
+{
+    type Response = ServiceResponse<BodyType>;
+    type Error = Error;
+    type Future = LocalBoxFuture<'static, Result<Self::Response, Self::Error>>;
+
+    forward_ready!(service);
+
+    // Asynchronously handle the request and bypass it to the next service
+    fn call(&self, req: ServiceRequest) -> Self::Future {
+        // Clone the service and validator to be able to use them in the async block
+        let service = Rc::clone(&self.service);
```

Comment:
This looks fine but you can consider changing this into an `Arc::clone()` instead due to it be thread-safe and atomic, I think it is slightly more costly than Rc but worth mentioning for scalability purposes

Reply:
Replaced `Rc` with `Arc`